sidechain/consensus/common/src/block_import.rs

```diff
@@ -56,22 +61,27 @@ where
 	where
 		F: FnOnce(Self::SidechainState) -> Result<Self::SidechainState, Error>;
 
-	/// key that is used for state encryption
+	/// Key that is used for state encryption.
 	fn state_key(&self) -> Self::StateCrypto;
 
-	/// import the block
-	fn import_block(
+	/// Import a sidechain block and mutate state by `apply_state_update`.
+	fn import_block<TopPoolExecutor>(
 		&self,
-		sidechain_block: SB,
+		signed_sidechain_block: SB,
 		parentchain_header: &PB::Header,
+		top_pool_executor: Arc<TopPoolExecutor>,
```

Comment:
Here we don't need to take the executor as `Arc<>`, because we're not taking ownership of it. It's enough to just use a borrow: `&TopPoolExecutor`.

Reply:
Hmm thinking about it, we can do this more elegantly: As it is right now, we have a lot of implementation in this trait, which IMO is not ideal. We should put all the logic of removing the trusted operation into the implementing struct:
- Add a new `fn cleanup(&SidechainBlock)` to this trait that is called in this implementation (or re-name to something better)
- In the implementing struct, implement the `cleanup()` by doing the check if we're the author and then removing from the top-pool. The implementing struct will then have the `TopPoolExecutor` as member (where we need an `Arc` again, because we have shared ownership.

Reply:
What @clangenb has started here, is also know as the [Template Method Design Pattern](https://www.geeksforgeeks.org/template-method-design-pattern/), where you have a recipe of steps that need to be done (implemented in the abstract class or trait in this case). And all the concrete steps are implemented in the concrete class/struct.

Reply:
I see. Thanks for the explanation. I think you resolved this issue yourself, is that correct @mullefel ?

Reply:
Yes indeed, resolving this conversation now ☺️